src/Cardano/Wallet/CoinSelection/LargestFirst.hs

```diff
@@ -0,0 +1,112 @@
+    => CoinSelectionOptions
+    -> UTxO
+    -> NonEmpty TxOut
+    -> ExceptT CoinSelectionError m CoinSelection
+largestFirst opt utxo txOutputs = do
+    let txOutputsSorted = NE.toList $ NE.sortBy (comparing coin) txOutputs
+    let n = fromIntegral $ maximumNumberOfInputs opt
+    let nLargest = take n . L.sortBy (flip $ comparing (coin . snd)) . Map.toList . getUTxO
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
+    let utxoBalance = fromIntegral $ balance utxo
+    let numberOfUtxoEntries = fromIntegral $ L.length $ (Map.toList . getUTxO) utxo
+    let numberOfTransactionOutputs = fromIntegral $ NE.length txOutputs
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ UtxoExhausted utxoBalance moneyRequested
+
+    -- FIXME ? we need to check if the transaction outputs are not redeemable
+
+    case foldM atLeast (L.reverse $ nLargest utxo, mempty) txOutputsSorted of
+        Just (_, s) ->
+            return s
+        Nothing ->
+            throwE $ MaximumInputsReached (fromIntegral n)
+
+
+{-------------------------------------------------------------------------------
+                       Helper types and functions
+-------------------------------------------------------------------------------}
+
+-- Select coins to cover at least the specified value
+-- When we fail in the random selection policy because we exceeded the maximum
+-- number of inputs @n@, we fallback on the 'largestFirstFallback'. We select
+-- the @n@ largest inputs from the UTxO in a single linear pass, then walk over
+-- these from large to small to try and cover the value we need to cover.
+-- If this fails, we have no further fallbacks and this payment request is
+-- not satisfiable.
+--
+-- If it succeeds, we can then use this as the basis for another call to
+-- the random input selection to try and construct a more useful change output
+-- (provided we haven't used up all available inputs yet).
+atLeast
+    :: ([(TxIn, TxOut)], CoinSelection)
+    -> TxOut
+    -> Maybe ([(TxIn, TxOut)], CoinSelection)
+atLeast (utxo0, selection) txout =
+    go (fromIntegral $ getCoin $ coin txout, mempty) utxo0
+    where
+    go :: (Int, [(TxIn, TxOut)])
+       -> [(TxIn, TxOut)]
+       -> Maybe ([(TxIn, TxOut)], CoinSelection)
+    go (target, ins) utxo
+        | target <= 0 = Just
+            ( utxo
+            , selection <> CoinSelection
+                { inputs = ins
+                , outputs = [txout]
+                , change = [Coin (fromIntegral $ abs target)]
+                }
+            )
+        | null utxo =
+            Nothing
+        | otherwise =
+            let
+                (inp, out):utxo' = utxo
+                target' =
+                    (fromIntegral (getCoin (coin txout)))
+                    - (fromIntegral (getCoin (coin out)))
```

Comment:
This seems wrong? The new target should be based on the current target and not on the initial one?

Reply:
we have `foldM atLeast utxos outputs` and in the `atLeast` we process every output per go. And inside, we are looking which first coin from utxo (previously taken the biggest maxnumberofinputs of them and then taken from the smallest) can cover this target. To my best knowledge, the target within go (I changed inner loop that searches for this to `pickTheFirst`) my stay intact and the above expression does its job.

Reply:
Hmmm... that isn't correct. The `go` may actually picks more than one UTxO (possibly depleting the whole list if there aren't enough to cover for the target), that's why it's an iterative approach.

The target starts being the txout's value. And then, after every pick, we makes it smaller by the amount of the corresponding UTxO entry we've picked, and iterate, until the target is covered. So, in the last branch, we should have:

```hs
targets' = target - fromIntegral (getCoin (coin out))
```

Reply:
ok, finally got it - corrected the algorithm and included boundary and exemplary cases in unit tests to show the implementation is right now